// @dagger.io/dagger/dist/introspector/scanner/abtractions/constructor.js
import { UnknownDaggerError } from "../../../common/errors/UnknownDaggerError.js";
import { Argument } from "./argument.js";
export class Constructor {
    checker;
    declaration;
    constructor(checker, declaration) {
        this.checker = checker;
        this.declaration = declaration;
    }
    get name() {
        return "";
    }
    get arguments() {
        return this.declaration.parameters.reduce((acc, param) => {
            const symbol = this.checker.getSymbolAtLocation(param.name);
            if (!symbol) {
                throw new UnknownDaggerError(`could not get constructor param: ${param.name.getText()}`, {});
            }
            const argument = new Argument(this.checker, symbol);
            acc[argument.name] = argument;
            return acc;
        }, {});
    }
    // TODO(TomChv): replace with `ToJson` method
    // after the refactor is complete.
    get typeDef() {
        return {
            args: Object.entries(this.arguments).reduce((acc, [name, arg]) => {
                acc[name] = arg.typeDef;
                return acc;
            }, {}),
        };
    }
    toJSON() {
        return {
            args: this.arguments,
        };
    }
    getArgOrder() {
        return Object.keys(this.arguments);
    }
}
